test/dag_benchmarks.cc

```diff
@@ -100,108 +100,159 @@ static constexpr instruction_graph_generator::policy_set benchmark_instruction_g
-	task_manager_benchmark_context& operator=(task_manager_benchmark_context&&) = delete;
-
-	~task_manager_benchmark_context() { tm.generate_epoch_task(celerity::detail::epoch_action::shutdown); }
+struct dag_benchmark_context {
+	dag_benchmark_context() = default;
+	virtual ~dag_benchmark_context() = 0;
+	CELERITY_DETAIL_UTILS_NON_COPYABLE(dag_benchmark_context);
+	CELERITY_DETAIL_UTILS_NON_MOVABLE(dag_benchmark_context);
 
 	template <int KernelDims, typename CGF>
 	void create_task(range<KernelDims> global_range, CGF cgf) {
-		tm.generate_command_group_task(invoke_command_group_function([=](handler& cgh) {
+		m_command_groups.push_back(invoke_command_group_function([=](handler& cgh) {
 			cgf(cgh);
-			cgh.host_task(global_range, [](partition<KernelDims>) {});
+			cgh.parallel_for(global_range, [](item<KernelDims>) {});
 		}));
 	}
+
+  protected:
+	std::vector<raw_command_group> m_command_groups;
 };
 
+dag_benchmark_context::~dag_benchmark_context() = default;
 
-struct command_graph_generator_benchmark_context : private task_manager::delegate { // NOLINT(cppcoreguidelines-virtual-class-destructor)
+struct tdag_benchmark_context : dag_benchmark_context, private task_manager::delegate {
 	const size_t num_nodes;
 	task_graph tdag;
 	task_recorder trec;
-	task_manager tm{num_nodes, tdag, test_utils::g_print_graphs ? &trec : nullptr, this, benchmark_task_manager_policy};
+	task_manager tm{num_nodes, tdag, test_utils::g_print_graphs ? &trec : nullptr, static_cast<task_manager::delegate*>(this), benchmark_task_manager_policy};
+	test_utils::mock_buffer_factory mbf{tm};
+
+	explicit tdag_benchmark_context(const size_t num_nodes = 1) : num_nodes(num_nodes) {}
+
+	void task_created(const task* tsk) override { m_tasks.push_back(tsk); }
+
+	void initialize() { tm.generate_epoch_task(celerity::detail::epoch_action::init); }
+
+	void prepare() { m_tasks.reserve(m_command_groups.size()); }
+
+	void execute() { create_all_tasks(); }
+
+	void finalize() {
+		tm.generate_epoch_task(celerity::detail::epoch_action::shutdown);
+		m_command_groups.clear();
+	}
+
+  protected:
+	std::vector<const task*> m_tasks; // for use in derived classes
+
+	void create_all_tasks() {
+		for(auto& cg : m_command_groups) {
+			tm.generate_command_group_task(std::move(cg));
+		}
+	}
+};
+
+struct cdag_benchmark_context : tdag_benchmark_context {
 	command_graph cdag;
 	command_recorder crec;
 	command_graph_generator cggen{num_nodes, 0 /* local_nid */, cdag, test_utils::g_print_graphs ? &crec : nullptr, benchmark_command_graph_generator_policy};
 	test_utils::mock_buffer_factory mbf{tm, cggen};
 
-	explicit command_graph_generator_benchmark_context(const size_t num_nodes) : num_nodes(num_nodes) { tm.generate_epoch_task(epoch_action::init); }
+	explicit cdag_benchmark_context(const size_t num_nodes) : tdag_benchmark_context(num_nodes) {}
+
+	void initialize() {
+		tdag_benchmark_context::initialize();
+		create_all_commands();
```

Comment:
This doesn't do anything since `m_tasks` is empty, right?

Also, what is it supposed to do, pre-allocate `m_command_batches`? Maybe also add a comment.